### PR TITLE
Add notification library

### PR DESCRIPTION
**Pull request by jakalkhalili**

### CODE REVIEW

File: internal/notifier/sender.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+// Listen is goroutine that listens for incoming messages and sends it
+func Listen(notification chan *Notification) {
+	n := <-notification
+	if len(n.Jenkins.Spec.Notification) > 0 {
+		for _, endpoint := range n.Jenkins.Spec.Notification {
+			var err error
+			var service Service
+			var selector v1alpha2.SecretKeySelector
+			secret := &corev1.Secret{}
+
```
Comment (tomaszsek) [inline internal/notifier/sender.go]: add notification filtering

File: internal/notifier/slack.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	if err != nil {
+		return err
+	}
+
+	resp, err := client.Do(request)
+	if err != nil {
+		return err
+	}
+
+	err = resp.Body.Close()
```
Comment (tomaszsek) [inline internal/notifier/slack.go]: defer

File: internal/notifier/slack_test.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
+package notifier
+
+import (
+	"encoding/json"
+	"github.com/stretchr/testify/assert"
```
Comment (tomaszsek) [inline internal/notifier/slack_test.go]: formatting

File: pkg/apis/jenkins/v1alpha2/jenkins_types.go [hunk 4]
```diff
@@ -17,6 +17,9 @@ type JenkinsSpec struct {
 	// +optional
 	SeedJobs []SeedJob `json:"seedJobs,omitempty"`
 
+	// Notification defines services which are used to inform about Jenkins behavior
+	Notification []Notification `json:"notification,omitempty"`
```
Comment (tomaszsek) [inline pkg/apis/jenkins/v1alpha2/jenkins_types.go]: Notification -> Notification, notification -> notifications

File: pkg/apis/jenkins/v1alpha2/jenkins_types.go [hunk 5]
```diff
@@ -17,6 +17,9 @@ type JenkinsSpec struct {
 	// +optional
 	SeedJobs []SeedJob `json:"seedJobs,omitempty"`
 
+	// Notification defines services which are used to inform about Jenkins behavior
```
Comment (tomaszsek) [inline pkg/apis/jenkins/v1alpha2/jenkins_types.go]: improve

File: pkg/apis/jenkins/v1alpha2/jenkins_types.go [hunk 6]
```diff
@@ -443,6 +484,20 @@ const (
 	UsernamePasswordCredentialType JenkinsCredentialType = "usernamePassword"
 )
 
+// JenkinsNotificationLogLevel defines type of Notification feature frequency of sending logger entries
+type JenkinsNotificationLogLevel string
+
+const (
+	// LogLevelNone - No logs
+	LogLevelNone JenkinsNotificationLogLevel = ""
```
Comment (tomaszsek) [inline pkg/apis/jenkins/v1alpha2/jenkins_types.go]: remove

File: internal/notifier/sender.go [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+	Namespace         string
+	CrName            string
+	LogLevel          LoggingLevel
+	Message           string
+	MessageVerbose    string
+}
+
+// Notification contains message which will be sent
+type Notification struct {
+	Jenkins     *v1alpha2.Jenkins
```
Comment (tomaszsek) [inline internal/notifier/sender.go]: don't use pointer

File: internal/notifier/sender.go [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+	K8sClient   k8sclient.Client
+	Logger      logr.Logger
+	Information *Information
```
Comment (tomaszsek) [inline internal/notifier/sender.go]: don't use pointer

File: internal/notifier/sender.go [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+				var err error
+				var svc service
+
+				if endpoint.Slack != (v1alpha2.Slack{}) {
+					svc = Slack{}
+				} else if endpoint.Teams != (v1alpha2.Teams{}) {
+					svc = Teams{}
+				} else if endpoint.Mailgun != (v1alpha2.Mailgun{}) {
+					svc = Mailgun{
+						Domain:    endpoint.Mailgun.Domain,
```
Comment (tomaszsek) [inline internal/notifier/sender.go]: remove

File: internal/notifier/sender.go [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+						Recipient: endpoint.Mailgun.Recipient,
```
Comment (tomaszsek) [inline internal/notifier/sender.go]: remove

File: internal/notifier/sender.go [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+						From:      endpoint.Mailgun.From,
```
Comment (tomaszsek) [inline internal/notifier/sender.go]: remove

File: internal/notifier/sender.go [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+					}
+				} else {
+					n.Logger.V(log.VWarn).Info("Notification service not found or not defined")
+				}
+
+				err = notify(svc, n)
```
Comment (tomaszsek) [inline internal/notifier/sender.go]: pass notificationConfig

File: internal/notifier/sender.go [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+	default:
+		return "#c8c8c8"
+	}
+}
+
+func notify(svc service, n *Notification) error {
+	var err error
+	switch s := svc.(type) {
+	case Slack:
+		err = s.Send(n)
```
Comment (tomaszsek) [inline internal/notifier/sender.go]: pass notificationConfig

File: internal/notifier/sender.go [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+	case Teams:
+		err = s.Send(n)
```
Comment (tomaszsek) [inline internal/notifier/sender.go]: pass notificationConfig

File: internal/notifier/sender.go [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+	case Mailgun:
+		err = s.Send(n)
```
Comment (tomaszsek) [inline internal/notifier/sender.go]: pass notificationConfig

File: internal/notifier/mailgun.go [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+	"github.com/jenkinsci/kubernetes-operator/pkg/log"
+
+	"github.com/mailgun/mailgun-go/v3"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// Mailgun is service for sending emails
+type Mailgun struct {
+	Domain    string
```
Comment (tomaszsek) [inline internal/notifier/mailgun.go]: remove

File: internal/notifier/mailgun.go [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+	Recipient string
```
Comment (tomaszsek) [inline internal/notifier/mailgun.go]: remove

File: internal/notifier/mailgun.go [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+	From      string
```
Comment (tomaszsek) [inline internal/notifier/mailgun.go]: remove

File: internal/notifier/mailgun.go [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+
+	i := n.Information
+
+	err := n.K8sClient.Get(context.TODO(), types.NamespacedName{Name: selector.Name, Namespace: n.Jenkins.Namespace}, secret)
+	if err != nil {
+		n.Logger.V(log.VWarn).Info(fmt.Sprintf("Failed to get secret with name `%s`. %+v", selector.Name, err))
+		return err
+	}
+
+	mg := mailgun.NewMailgun(m.Domain, secret.StringData[selector.Name])
```
Comment (tomaszsek) [inline internal/notifier/mailgun.go]: `secret.StringData` will be empty use `secret.Data`

Comment (tomaszsek) [reply internal/notifier/mailgun.go]: check if value is empty

File: internal/notifier/mailgun.go [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+
+	content := `
```
Comment (tomaszsek) [inline internal/notifier/mailgun.go]: move out from function and declare as const

File: internal/notifier/msteams.go [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+	"github.com/jenkinsci/kubernetes-operator/pkg/apis/jenkins/v1alpha2"
+	"github.com/jenkinsci/kubernetes-operator/pkg/log"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// Teams is Microsoft Teams Service
+type Teams struct {
+	apiURL string
```
Comment (tomaszsek) [inline internal/notifier/msteams.go]: remove

File: internal/notifier/mailgun.go [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+// Mailgun is service for sending emails
+type Mailgun struct {
+	Domain    string
+	Recipient string
+	From      string
+}
+
+// Send is function for sending directly to API
+func (m Mailgun) Send(n *Notification) error {
+	var selector v1alpha2.SecretKeySelector
```
Comment (tomaszsek) [inline internal/notifier/mailgun.go]: value not set

File: internal/notifier/msteams.go [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+
+// TeamsFact is field where we can put content
+type TeamsFact struct {
+	Name  string `json:"name"`
+	Value string `json:"value"`
+}
+
+// Send is function for sending directly to API
+func (t Teams) Send(n *Notification) error {
+	var selector v1alpha2.SecretKeySelector
```
Comment (tomaszsek) [inline internal/notifier/msteams.go]: value not set

File: internal/notifier/msteams.go [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+						Name:  namespaceFieldName,
+						Value: i.Namespace,
+					},
+				},
+				Text: i.Message,
+			},
+		},
+	})
+
+	if t.apiURL == "" {
```
Comment (tomaszsek) [inline internal/notifier/msteams.go]: remove condition

File: internal/notifier/msteams.go [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+		err := n.K8sClient.Get(context.TODO(), types.NamespacedName{Name: selector.Name, Namespace: n.Jenkins.Namespace}, secret)
+		if err != nil {
+			n.Logger.V(log.VWarn).Info(fmt.Sprintf("Failed to get secret with name `%s`. %+v", selector.Name, err))
+		}
+
+		t.apiURL = secret.StringData[selector.Name]
```
Comment (tomaszsek) [inline internal/notifier/msteams.go]: use `secret.Data` and check if value is empty

File: internal/notifier/msteams_test.go [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+				}
+			case crNameFieldName:
+				if fact.Value != i.CrName {
+					t.Fatalf("%s is not equal! Must be %s", crNameFieldName, i.CrName)
+				}
+			case statusFieldName:
+				if fact.Value != getStatusName(i.Status) {
+					t.Fatalf("%s is not equal! Must be %s", statusFieldName, getStatusName(i.Status))
+				}
+			}
```
[the remark this answers is not in the document]
Comment (tomaszsek) [reply internal/notifier/msteams_test.go]: ???

File: internal/notifier/sender.go [hunk 27]
```diff
@@ -0,0 +1,179 @@
+package notifier
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/go-logr/logr"
```
[the remark this answers is not in the document]
Comment (tomaszsek) [reply internal/notifier/sender.go]: ???

File: internal/notifier/slack.go [hunk 28]
```diff
@@ -0,0 +1,116 @@
...
+// SlackField is representation of json field.
+type SlackField struct {
+	Title string `json:"title"`
+	Value string `json:"value"`
+	Short bool   `json:"short"`
+}
+
+// Send is function for sending directly to API
+func (s Slack) Send(n *Notification) error {
+	var selector v1alpha2.SecretKeySelector
```
Comment (tomaszsek) [inline internal/notifier/slack.go]: value not set

File: internal/notifier/slack.go [hunk 29]
```diff
@@ -0,0 +1,116 @@
...
+	secret := &corev1.Secret{}
+
+	i := n.Information
+
+	if s.apiURL == "" {
```
Comment (tomaszsek) [inline internal/notifier/slack.go]: remove condition

File: internal/notifier/slack.go [hunk 30]
```diff
@@ -0,0 +1,116 @@
...
+		err := n.K8sClient.Get(context.TODO(), types.NamespacedName{Name: selector.Name, Namespace: n.Jenkins.Namespace}, secret)
+		if err != nil {
+			n.Logger.V(log.VWarn).Info(fmt.Sprintf("Failed to get secret with name `%s`. %+v", selector.Name, err))
+		}
+
+		s.apiURL = secret.StringData[selector.Name]
```
Comment (tomaszsek) [inline internal/notifier/slack.go]: use `secret.Data` and check if it's empty

File: internal/notifier/slack_test.go [hunk 31]
```diff
@@ -0,0 +1,61 @@
...
+			switch field.Title {
+			case configurationTypeFieldName:
+				assert.Equal(t, field.Value, i.ConfigurationType)
+			case crNameFieldName:
+				assert.Equal(t, field.Value, i.CrName)
+			case messageFieldName:
+				assert.Equal(t, field.Value, i.Message)
+			case loggingLevelFieldName:
+				assert.Equal(t, field.Value, string(i.LogLevel))
+			}
```
Comment (tomaszsek) [inline internal/notifier/slack_test.go]: fail in default case

File: pkg/apis/jenkins/v1alpha2/jenkins_types.go [hunk 32]
```diff
@@ -50,6 +53,44 @@ type JenkinsSpec struct {
...
+	Verbose      bool                        `json:"verbose"`
+	Name         string                      `json:"name"`
+	Slack        Slack                       `json:"slack,omitempty"`
+	Teams        Teams                       `json:"teams,omitempty"`
+	Mailgun      Mailgun                     `json:"mailgun,omitempty"`
+}
+
+// Slack is handler for Slack
+type Slack struct {
+	// The web hook url to Slack App
```
Comment (tomaszsek) [inline pkg/apis/jenkins/v1alpha2/jenkins_types.go]: url -> URL

File: pkg/apis/jenkins/v1alpha2/jenkins_types.go [hunk 33]
```diff
@@ -50,6 +53,44 @@ type JenkinsSpec struct {
...
+	URLSecretKeySelector SecretKeySelector `json:"urlSecretKeySelector"`
+}
+
+// Teams is handler for Microsoft Teams
+type Teams struct {
+	// The web hook url to Teams App
```
Comment (tomaszsek) [inline pkg/apis/jenkins/v1alpha2/jenkins_types.go]: url -> URL

File: internal/notifier/msteams_test.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (jakalkhalili) [reply internal/notifier/msteams_test.go]: done

File: internal/notifier/sender.go [hunk 34]
```diff
@@ -0,0 +1,151 @@
...
+type service interface {
+	Send(i *Notification) error
+}
+
+// Listen is goroutine that listens for incoming messages and sends it
+func Listen(notification chan *Notification) {
+	<-notification
+	for n := range notification {
+		if len(n.Jenkins.Spec.Notification) > 0 {
+			for _, endpoint := range n.Jenkins.Spec.Notification {
```
[the remark this answers is not in the document]
Comment (jakalkhalili) [reply internal/notifier/sender.go]: done

File: pkg/apis/jenkins/v1alpha2/jenkins_types.go [hunk 35]
```diff
@@ -17,6 +17,10 @@ type JenkinsSpec struct {
 	// +optional
 	SeedJobs []SeedJob `json:"seedJobs,omitempty"`
 
+	// Notification defines services which are used to inform about Jenkins behavior
+	// Can be used to integrate chat services like Slack or Email services like Mailgun
+	Notification Notification `json:"notifications,omitempty"`
```
Comment (tomaszsek) [inline pkg/apis/jenkins/v1alpha2/jenkins_types.go]: Notifications []Notification `json:"notifications,omitempty"`

File: pkg/apis/jenkins/v1alpha2/jenkins_types.go [hunk 36]
```diff
@@ -17,6 +17,10 @@ type JenkinsSpec struct {
 	// +optional
 	SeedJobs []SeedJob `json:"seedJobs,omitempty"`
 
+	// Notification defines services which are used to inform about Jenkins behavior
```
Comment (tomaszsek) [inline pkg/apis/jenkins/v1alpha2/jenkins_types.go]: // Notifications defines services which are used to inform about Jenkins status

File: internal/notifier/mailgun.go [hunk 37]
```diff
@@ -0,0 +1,76 @@
...
+		<h6 style="font-size: 11px; color: grey; margin-top: 15px;">Powered by Jenkins Operator <3</h6>
+</body>
+</html>`
+
+// Mailgun is service for sending emails
+type Mailgun struct{}
+
+// Send is function for sending directly to API
+func (m Mailgun) Send(n *Notification, config v1alpha2.Notification) error {
+	var selector v1alpha2.SecretKeySelector
```
Comment (tomaszsek) [inline internal/notifier/mailgun.go]: remove

File: internal/notifier/mailgun.go [hunk 38]
```diff
@@ -0,0 +1,76 @@
...
+	secret := &corev1.Secret{}
+	i := n.Information
+
+	selector = config.Mailgun.APIKeySecretKeySelector
```
Comment (tomaszsek) [inline internal/notifier/mailgun.go]: selector := config.Mailgun.APIKeySecretKeySelector

File: internal/notifier/mailgun.go [hunk 39]
```diff
@@ -0,0 +1,76 @@
...
+
+	err := n.K8sClient.Get(context.TODO(), types.NamespacedName{Name: selector.Name, Namespace: n.Jenkins.Namespace}, secret)
+	if err != nil {
+		n.Logger.V(log.VWarn).Info(fmt.Sprintf("Failed to get secret with name `%s`. %+v", selector.Name, err))
```
Comment (tomaszsek) [inline internal/notifier/mailgun.go]: return error instead

File: internal/notifier/mailgun.go [hunk 40]
```diff
@@ -0,0 +1,76 @@
...
+		return err
+	}
+
+	secretValue := string(secret.Data[selector.Name])
+	if secretValue == "" {
+		return fmt.Errorf("SecretValue %s is empty", selector.Name)
```
Comment (tomaszsek) [inline internal/notifier/mailgun.go]: return errors.Errorf("SecretValue %s is empty", selector.Name)

File: internal/notifier/msteams.go (hunk 23 above)
Comment (tomaszsek) [reply internal/notifier/msteams.go]: delete

File: internal/notifier/msteams.go [hunk 41]
```diff
@@ -0,0 +1,106 @@
...
+	Value string `json:"value"`
+}
+
+// Send is function for sending directly to API
+func (t Teams) Send(n *Notification, config v1alpha2.Notification) error {
+	var selector v1alpha2.SecretKeySelector
+	secret := &corev1.Secret{}
+	i := n.Information
+
+	selector = config.Teams.URLSecretKeySelector
```
Comment (tomaszsek) [inline internal/notifier/msteams.go]: selector := config.Teams.URLSecretKeySelector

File: internal/notifier/msteams.go [hunk 42]
```diff
@@ -0,0 +1,106 @@
...
+
+	err := n.K8sClient.Get(context.TODO(), types.NamespacedName{Name: selector.Name, Namespace: n.Jenkins.Namespace}, secret)
+	if err != nil {
+		n.Logger.V(log.VWarn).Info(fmt.Sprintf("Failed to get secret with name `%s`. %+v", selector.Name, err))
```
Comment (tomaszsek) [inline internal/notifier/msteams.go]: return err instead

File: internal/notifier/msteams.go [hunk 43]
```diff
@@ -0,0 +1,106 @@
...
+					},
+				},
+				Text: i.Message,
+			},
+		},
+	})
+
+	secretValue := string(secret.Data[selector.Key])
+	if secretValue == "" {
+		return fmt.Errorf("SecretValue %s is empty", selector.Name)
```
Comment (tomaszsek) [inline internal/notifier/msteams.go]: return errors.Errorf("SecretValue %s is empty", selector.Name)

File: internal/notifier/sender.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (tomaszsek) [reply internal/notifier/sender.go]: ???

File: internal/notifier/sender.go [hunk 44]
```diff
@@ -0,0 +1,144 @@
...
+			return "gray"
+		}
+	default:
+		return "#c8c8c8"
+	}
+}
+
+func notify(svc service, n *Notification, nc v1alpha2.Notification) error {
+	var err error
+	switch s := svc.(type) {
```
Comment (tomaszsek) [inline internal/notifier/sender.go]: remove entire switch

File: internal/notifier/sender.go [hunk 45]
```diff
@@ -0,0 +1,144 @@
...
+	case Slack:
+		err = s.Send(n, nc)
+	case Teams:
+		err = s.Send(n, nc)
+	case Mailgun:
+		err = s.Send(n, nc)
+	}
+
+	return err
```
Comment (tomaszsek) [inline internal/notifier/sender.go]: return s.Send(n, nc)

File: internal/notifier/slack.go (hunk 28 above)
Comment (tomaszsek) [reply internal/notifier/slack.go]: remove

File: internal/notifier/slack.go [hunk 46]
```diff
@@ -0,0 +1,116 @@
...
+	Short bool   `json:"short"`
+}
+
+// Send is function for sending directly to API
+func (s Slack) Send(n *Notification, config v1alpha2.Notification) error {
+	var selector v1alpha2.SecretKeySelector
+	secret := &corev1.Secret{}
+	i := n.Information
+
+	selector = config.Slack.URLSecretKeySelector
```
Comment (tomaszsek) [inline internal/notifier/slack.go]: selector := config.Slack.URLSecretKeySelector

File: internal/notifier/slack.go [hunk 47]
```diff
@@ -0,0 +1,116 @@
...
+
+	err := n.K8sClient.Get(context.TODO(), types.NamespacedName{Name: selector.Name, Namespace: n.Jenkins.Namespace}, secret)
+	if err != nil {
+		n.Logger.V(log.VWarn).Info(fmt.Sprintf("Failed to get secret with name `%s`. %+v", selector.Name, err))
```
Comment (tomaszsek) [inline internal/notifier/slack.go]: return err instead

File: internal/notifier/slack.go [hunk 48]
```diff
@@ -0,0 +1,116 @@
...
+					},
+				},
+				Footer: footerContent,
+			},
+		},
+	})
+
+	secretValue := string(secret.Data[selector.Key])
+	if secretValue == "" {
+		return fmt.Errorf("SecretValue %s is empty", selector.Name)
```
Comment (tomaszsek) [inline internal/notifier/slack.go]: return errors.Errorf("SecretValue %s is empty", selector.Name)

File: pkg/apis/jenkins/v1alpha2/jenkins_types.go (hunk 32 above)
Comment (tomaszsek) [reply pkg/apis/jenkins/v1alpha2/jenkins_types.go]: ???

File: pkg/apis/jenkins/v1alpha2/jenkins_types.go (hunk 33 above)
Comment (tomaszsek) [reply pkg/apis/jenkins/v1alpha2/jenkins_types.go]: ???

File: internal/notifier/mailgun.go [hunk 49]
```diff
@@ -0,0 +1,73 @@
+package notifier
+
+import (
+	"context"
+	"fmt"
+	"github.com/pkg/errors"
```
Comment (tomaszsek) [inline internal/notifier/mailgun.go]: formatting

File: internal/notifier/msteams_test.go [hunk 50]
```diff
@@ -0,0 +1,98 @@
...
+
+		Data: map[string][]byte{
+			testURLSelectorKeyName: []byte(server.URL),
+		},
+	}
+
+	err := notification.K8sClient.Create(context.TODO(), secret)
+	assert.NoError(t, err)
+
+	assert.NoError(t, teams.Send(notification, v1alpha2.Notification{
```
Comment (tomaszsek) [inline internal/notifier/msteams_test.go]: assign to err then check

File: internal/notifier/sender.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (tomaszsek) [reply internal/notifier/sender.go]: ???

File: internal/notifier/sender.go [hunk 51]
```diff
@@ -0,0 +1,141 @@
...
+
+// Service is skeleton for additional services
+type service interface {
+	Send(i *Notification, config v1alpha2.Notification) error
+}
+
+// Listen is goroutine that listens for incoming messages and sends it
+func Listen(notification chan *Notification) {
+	for n := range notification {
+		if len(n.Jenkins.Spec.Notifications) > 0 {
```
Comment (tomaszsek) [inline internal/notifier/sender.go]: remove

File: internal/notifier/slack.go [hunk 52]
```diff
@@ -0,0 +1,113 @@
+package notifier
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"github.com/pkg/errors"
```
Comment (tomaszsek) [inline internal/notifier/slack.go]: formatting

File: internal/notifier/slack_test.go [hunk 53]
```diff
@@ -0,0 +1,97 @@
...
+			testURLSelectorKeyName: []byte(server.URL),
+		},
+	}
+
+	err := notification.K8sClient.Create(context.TODO(), secret)
+	assert.NoError(t, err)
+
+	slack := Slack{}
+
+	assert.NoError(t, slack.Send(notification, v1alpha2.Notification{
```
Comment (tomaszsek) [inline internal/notifier/slack_test.go]: assign to err then check

**Pull request merged**